### ISSUE TITLE
Debugger is not able to step into non-inlined hardware intrinsics

### ISSUE THREAD

**Issue by jkotas** [dotnet/runtime#61890]
Repro:
```
using System.Runtime.Intrinsics;
using System.Runtime.Intrinsics.X86;

byte mask = (byte)Random.Shared.Next(10);
var m = Avx2.AlignRight(default(Vector256<byte>), default(Vector256<byte>), mask);
```

Run this under VS debugger and try to step into `Avx2.AlignRight`.

Expected result: Able to step into `Avx2.AlignRight` and see its code.
Actual result: Unable to step into `Avx2.AlignRight`. I have tried disabling "Just My Code", going to disassembly window and stepping there, ... nothing works.

**Comment by jkotas** [comment]
See https://github.com/dotnet/runtime/issues/61877#issuecomment-974768007 for full context.

**Comment by tannergooding** [comment]
This is expected AFAIK. Most intrinsics synthesize trees behind the scenes and so there is no user code to actually step into.

In the case of something like `Avx2.AlignRight` (and other hardware intrinsics) the C# implementation is simply a recursive call to itself and all the interesting logic (the up to 256 statement jump table) is generated entirely during codegen.

This is done to save space, reduce errors, and simplify the implementation overall.

--------------------------

Now, I don't know why you can't single step into https://github.com/dotnet/runtime/blob/69b5d67d9418d672609aa6e2c418a3d4ae00ad18/src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/X86/Avx2.cs#L105-L109

Perhaps this has something to do with R2R?

**Comment by jkotas** [comment]
> Now, I don't know why you can't single step into the recursive call

Right. And you should be able to see the disassembly that was the key part preventing from self-diagnosing #61877. 

> Perhaps this has something to do with R2R?

Still does not work, even with R2R disabled.

**Comment by zvrba** [comment]
Yes, being able to see disassembly would have helped immensly. But now that I've thought a bit more about it, I've come to think that codegen is behaving unreasonably: I asked it to emit a particular instruction, but it couldn't and it emitted snail-slow code instead. What it _should_ have done, IMO, is emit code that throws `InvalidProgramException` with a reasonable message.

If one _still_ wanted CLR to emit "whatever helper code" is necessary when direct translation is impossible, there could exist an "intrinsic" in a sub-namespace say `Emulated` , so `Avx2.Emulated.AlignRight`.

@tannergooding Doesn't codegen know about semantics about VPALIGRN? It could emit only a 16-element jump table. Higher values of the immediate byte always result in zero result.

**Comment by tannergooding** [comment]
>  I asked it to emit a particular instruction

And it did. It emitted `palignr` and generated the correct code to handle what you gave as the input.

> but it couldn't and it emitted snail-slow code instead.

The hardware intrinsics emit particular instructions. There is no guarantee on how fast or slow they will be (`CPUID` is always expensive, `PEXT/PDEP` are really expensive on certain hardware, etc), just that some instruction will be emitted to handle a given method.

There may be surrounding code emitted as part of that to ensure results end up in the target register if the instruction is "Read-Modify-Write". Folding may be done in other cases to allow something more efficient and remove the need for an explicit load/store.

In the case of values that require constant inputs, they may emit a call to a jump table to ensure the relevant instruction is emitted instead.

As with any performance oriented code, you should profile and measure the impact of the code you write for every platform/target you care about.

> What it should have done, IMO, is emit code that throws InvalidProgramException with a reasonable message.

That leads to otherwise "standard" .NET behavior not working. It means these instructions can't be used indirectly (via function pointers/delegates). It means they can't reliably be used in contexts where something becomes a constant after inlining and so will be constant in "hot code", etc.

We made the decision that intrinsics would work just like normal methods and would be optimized where-possible to be inline instructions. This results in the right thing happening the majority of the time, but in some edge cases or in debug code may result in functioning code that runs a bit slower.
